Make tf2_py compatible with python3.

de-vri-es opened the pull request:
This PR makes tf2_py compatible with python3. I compiled the package for both python 3.5 and 2.7 (which worked), but I **did not** test the resulting python module.

There were three issues to fix:
- Python3 uses unicode strings for text. The old python2 string type no longer exists.
- Module initialization has been overhauled for python3.
- Type definitions changed slightly.

I put wrapper functions for the correct string types in `python_compat.h` (and one convenience function for importing modules). Something similar could be done for the type definition and the module initialization, but those would involve writing new macros, which would not necessarily make the code easier to maintain (especially in the case of module initialization).

I did not use `py3c` (a C API compatibility layer for python2/3) to avoid adding a new dependency. If a new dependency is acceptable, `py3c` would be a good alternative. It would involve adding a new rosdep key though, I assume.

If this is the way to go, `PyObject_BorrowAttrString` could also be moved to `python_compat.h`, to group python convenience functions together. (Though I would rename it so it doesn't look like a real python function. It took me a while of digging through the python C API before realising it is defined in the source file itself).

de-vri-es committed:
tf2_py: Use PyUnicode objects for text in python3.

de-vri-es committed:
tf2_py: Make module initialization python3 compatible.

de-vri-es committed:
tf2_py: Fix type definition for python3.

de-vri-es commented:
This should fix #50, btw.

de-vri-es commented:
One more addendum: Rewriting the whole thing in Cython seems like a good idea to me, perhabs even with a more full-featured API (while maintaining backwards compatibility of course).

However, I wanted python3 compatibility fast, and this was the quickest, least intrusive way to get that.

tfoote commented:
@ros-pull-request-builder retest this please

Thanks for the contribution. I'll try to review it tomorrow. Python3 support will be great.

tfoote commented:
Thanks for taking the time to do this. It looks good to me. Bumping `PyObject_BorrowAttrString` into the `python_compat.h` makes sense to me. 

Thanks for working around it without the extra dependency. If this is enough lets stick with this and not completely refactor it to Cython.

de-vri-es committed:
tf2_py: Move and rename PyObject_BorrowAttrString.

de-vri-es commented:
Moved `PyObject_BorrowAtrrString` to `python_compat.h` and renamed it to `pythonBorrowAttrString` to avoid looking like it is part of the official Python API.

(Update: And made it inline, since it is defined in a header.)

Thanks for the review.

tfoote commented:
Thanks again!

spaghetti- commented:
Thank you for this patch!

koniarik commented:
Hi folks, will there be anytime soon release of geometry2 with  this patch?

nstiurca commented:
@SquirrelCZE I'm wondering the same thing.

In the meantime, I built from source based on the latest version I just pulled (439e2350ca9b2f0581a3631f0929acc2021db73e), but I can't figure out how to actually use TF2 in Python 3. Here's what I did:

```
mkdir -p overlay_ws/src
cd overlay_ws/src
git clone https://github.com/ros/geometry2.git
catkin_init_workspace
cd ..
rosdep install --from-paths src --os=ubuntu:xenial
catkin_make --use-ninja --cmake-args -DCMAKE_BUILD_TYPE=Release
source devel/setup.bash
```

Then I tried to use it with

```
$ python3
Python 3.5.2 (default, Sep 10 2016, 08:21:44) 
[GCC 5.4.0 20160609] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import tf2_py
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/nicu/dbot/overlay_ws/devel/lib/python2.7/dist-packages/tf2_py/__init__.py", line 35, in <module>
    exec(__fh.read())
  File "<string>", line 37, in <module>
ImportError: No module named '_tf2'
>>> 
```

nstiurca commented:
I should say the import error is the same (except for paths) I get when only working with the latest version installed from APT.

MohamadHalwani commented:
> @SquirrelCZE I'm wondering the same thing.
> 
> In the meantime, I built from source based on the latest version I just pulled ([439e235](https://github.com/ros/geometry2/commit/439e2350ca9b2f0581a3631f0929acc2021db73e)), but I can't figure out how to actually use TF2 in Python 3. Here's what I did:
> 
> ```
> mkdir -p overlay_ws/src
> cd overlay_ws/src
> git clone https://github.com/ros/geometry2.git
> catkin_init_workspace
> cd ..
> rosdep install --from-paths src --os=ubuntu:xenial
> catkin_make --use-ninja --cmake-args -DCMAKE_BUILD_TYPE=Release
> source devel/setup.bash
> ```
> 
> Then I tried to use it with
> 
> ```
> $ python3
> Python 3.5.2 (default, Sep 10 2016, 08:21:44) 
> [GCC 5.4.0 20160609] on linux
> Type "help", "copyright", "credits" or "license" for more information.
> >>> import tf2_py
> Traceback (most recent call last):
>   File "<stdin>", line 1, in <module>
>   File "/home/nicu/dbot/overlay_ws/devel/lib/python2.7/dist-packages/tf2_py/__init__.py", line 35, in <module>
>     exec(__fh.read())
>   File "<string>", line 37, in <module>
> ImportError: No module named '_tf2'
> >>> 
> ```

Have you solved this ??? Coming from the future :D
